### PR TITLE
feat: add appointments and clinic availability

### PR DESCRIPTION
**Pull request by DanielDervishi**
The goal of this pr is to provide secretaries the ability to schedule non-conflicting appointments between doctors and patients within a frame of availability provided by a clinic. This is achieved by:

- Adding and storing weekly re-occurring hours of operation (also known as availability) under the clinic that is mutable by an admin and is used to determine when appointments can be booked.
- Adding the ability for secretaries to book appointments between a doctor and patient on a given day, provided that they do not overlap with another appointment and that they fall within the clinic's operating hours (availability).

### CODE REVIEW

File: src/entities/UniversalTimeBlock.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package entities;
+
+import java.time.DayOfWeek;
+import java.time.LocalTime;
+
+public interface UniversalTimeBlock {
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlock.java:10]: Interface is missing javadocs

File: src/entities/UniversalTimeBlockWithDay.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package entities;
+
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+
+public interface UniversalTimeBlockWithDay extends UniversalTimeBlock {
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlockWithDay.java:9]: Interface missing javadocs

File: src/entities/UniversalTimeBlock.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+    LocalTime startTime();
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlock.java]: Method is missing javadocs. Furthermore, method should be named `getStartTime`. What happened to java conventions?

File: src/entities/UniversalTimeBlock.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    LocalTime endTime();
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlock.java]: Method is is missing javadocs. Furthermore, method should be named `getEndTime`. What happened to java conventions?

File: src/entities/UniversalTimeBlock.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    DayOfWeek dayOfWeek();
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlock.java]: Method is missing javadocs. Furthermore, method should be named `getDayOfWeek`. What happened to java conventions?

File: src/entities/UniversalTimeBlockWithDay.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    LocalDate date();
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlockWithDay.java]: Method is missing javadocs

File: src/entities/UniversalTimeBlockWithDay.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+
+    @Override
+    default DayOfWeek dayOfWeek(){
```
Comment (ahmedkhalf) [inline src/entities/UniversalTimeBlockWithDay.java]: Method is missing  `{@inheritDoc}` javadoc

File: src/entities/TimeBlock.java [hunk 8]
```diff
@@ -2,19 +2,18 @@
```
Comment (ahmedkhalf) [inline src/entities/TimeBlock.java]: Remove this comment

File: src/entities/TimeBlock.java [hunk 9]
```diff
@@ -72,5 +71,24 @@ public LocalTime endTimeToLocal(){
...
+    public LocalDate date() {
+        return startTime.toLocalDate();
+    }
+    @Override
+    public DayOfWeek dayOfWeek() {
+        return startTime.getDayOfWeek();
+    }
 }
 
+
```
Comment (ahmedkhalf) [inline src/entities/TimeBlock.java:77]: Too many blank lines at the end of the file. (max: 1).

File: src/entities/Clinic.java [hunk 10]
```diff
@@ -103,8 +104,13 @@ public ArrayList<Availability> getClinicHours() {
      * Sets the clinic's hours of operation.
      * @param clinicHours TimeBlock representing the clinic's new hours of operation.
      */
-    public void addClinicHours(Availability clinicHours) {
-        this.clinicHours.add(clinicHours);
+    public void addClinicHours(Availability clinicHours) {this.clinicHours.add(clinicHours);}
```
Comment (ahmedkhalf) [inline src/entities/Clinic.java]: Expand this method onto three lines.

File: src/entities/Availability.java [hunk 11]
```diff
@@ -8,7 +8,7 @@
 /**
```
Comment (ahmedkhalf) [inline src/entities/Availability.java:6]: Remove comment saying `PHASE 2 FILE` at the start of this file.

File: src/entities/Appointment.java [hunk 12]
```diff
@@ -4,10 +4,13 @@
```
Comment (ahmedkhalf) [inline src/entities/Appointment.java:2]: Remove comment saying `PHASE 2 FILE` at the start of this file.

File: src/dataBundles/TimeBlockData.java [hunk 13]
```diff
@@ -3,14 +3,16 @@
 package dataBundles;
```
Comment (ahmedkhalf) [inline src/dataBundles/TimeBlockData.java:1]: Remove comment saying `PHASE 2 FILE` at the start of this file.

File: src/entities/TimeBlock.java [hunk 14]
```diff
@@ -72,5 +71,24 @@ public LocalTime endTimeToLocal(){
         return startTime.toLocalTime();
     }
```
Comment (ahmedkhalf) [inline src/entities/TimeBlock.java]: Useless methods. Already local time. Please remove the `endTimeToLocal`, `startTimeToLocal` methods.

File: src/dataBundles/TimeBlockData.java [hunk 15]
```diff
@@ -44,5 +46,19 @@ public LocalTime endTimeToLocal(){
         return timeBlock.endTimeToLocal();
     }
```
Comment (ahmedkhalf) [inline src/dataBundles/TimeBlockData.java:37]: Useless methods. Already local time. Please remove the `endTimeToLocal` and `startTimeToLocal` methods.

File: src/entities/Availability.java [hunk 16]
```diff
@@ -63,4 +63,18 @@ public void setDoctorEndTime(LocalTime endTime) {
         this.doctorEndTime = endTime;
     }
```
Comment (ahmedkhalf) [inline src/entities/Availability.java:42]: Useless methods. `getDoctorEndTime` and `getDoctorStartTime` do the same thing as `getStartTime` and `getEndTime`.

File: src/dataBundles/AvailabilityData.java [hunk 17]
```diff
@@ -42,4 +43,18 @@ public LocalTime getDoctorEndTime() {
         return availability.getDoctorEndTime();
     }
```
Comment (ahmedkhalf) [inline src/dataBundles/AvailabilityData.java]: Useless methods. `getDoctorStartTime` and `getDoctorEndTime` do the same thing as  `getStartTime` and `getEndTime`.

File: src/dataBundles/AvailabilityData.java [hunk 18]
```diff
@@ -3,14 +3,15 @@
 package dataBundles;
```
Comment (ahmedkhalf) [inline src/dataBundles/AvailabilityData.java:1]: Remove `PHASE 2 FILE` comment at the start of the file.

File: src/dataBundles/AppointmentData.java [hunk 19]
```diff
@@ -3,11 +3,15 @@
 package dataBundles;
```
Comment (ahmedkhalf) [inline src/dataBundles/AppointmentData.java:1]: Remove `PHASE 2 FILE` comment at the start of the file.

File: src/presenters/entityViews/AppointmentView.java [hunk 20]
```diff
@@ -3,7 +3,10 @@
 package presenters.entityViews;
```
Comment (ahmedkhalf) [inline src/presenters/entityViews/AppointmentView.java:1]: Remove `PHASE 2 FILE` comment at the start of this file.

File: src/presenters/entityViews/AvailabilityView.java [hunk 21]
```diff
@@ -3,45 +3,44 @@
 package presenters.entityViews;
```
Comment (ahmedkhalf) [inline src/presenters/entityViews/AvailabilityView.java:1]: Remove `PHASE 2 FILE` comment at the start of this file.

File: src/presenters/entityViews/AvailabilityView.java [hunk 22]
```diff
@@ -3,45 +3,44 @@
...
     /**
      * @param item AvailabilityData  to view.
      * @return String representing item's full availability view.
      */
     @Override
     public String viewFull(AvailabilityData item) {
-        return viewDayOfWeek(item)
-                + viewDoctorStartTime(item)
-                + viewDoctorEndTime(item);
+        return dayOfWeekUtils.dayOfWeekToString(item.getDayOfWeek()) + ": " + item.startTime() + " to " + item.endTime();
```
Comment (ahmedkhalf) [inline src/presenters/entityViews/AvailabilityView.java]: Line exceeds 120 characters.

File: src/presenters/entityViews/ClinicView.java [hunk 23]
```diff
@@ -64,20 +67,31 @@ public String viewPhoneNumber(ClinicData item) {
      * @param item ClinicData to view.
      * @return String representing the clinic's hours of operation as a view.
      */
-    public String viewClinicHours(ClinicData item) {
-        ArrayList<String> al = new ArrayList<>();
-        for (int i = 0; i <= 6; i++) {
-            al.add(item.getClinicHours().get(i).getDoctorStartTime().toString());
-            al.add(item.getClinicHours().get(i).getDoctorEndTime().toString());
+
+    public String viewClinicHours(ClinicData item){
```
Comment (ahmedkhalf) [inline src/presenters/entityViews/ClinicView.java:68]: This method is too large. The `LinkedHashMap` should be returned from the `DayOfWeekUtils` file and take one line to create in this file.

File: src/presenters/response/AbsenceDetails.java [hunk 24]
```diff
@@ -0,0 +1,22 @@
+package presenters.response;
+
+import utilities.TimeUtils;
+
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+
+//public class AbsenceDetails {
```
Comment (ahmedkhalf) [inline src/presenters/response/AbsenceDetails.java]: This file should be removed if it's not going to be used.

File: src/presenters/response/AvailabilityDetails.java [hunk 25]
```diff
@@ -0,0 +1,28 @@
+package presenters.response;
+
+import java.time.DayOfWeek;
+import java.time.LocalTime;
+
+public class AvailabilityDetails {
```
Comment (ahmedkhalf) [inline src/presenters/response/AvailabilityDetails.java]: This is a not a correct response, please refer to the `README.md` file of this submodule for how to create a correct response class automatically using java record.

File: src/presenters/screenViews/ClinicScreenView.java [hunk 26]
```diff
@@ -90,4 +103,78 @@ public void showSuccessfullyChangedClinicPhoneNumber() {
...
+     */
+    public void showSuccessfullyDeletedAvailability(){
+        successMessage("Availability removed successfully.");
+    }
+
+    /**
+     * Shows the clinic's hours throughout the week given the clinic data.
+     * @param clinicData
+     */
+
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/ClinicScreenView.java]: Remove that empty line between javadoc and method.

File: src/presenters/screenViews/ClinicScreenView.java [hunk 27]
```diff
@@ -90,4 +103,78 @@ public void showSuccessfullyChangedClinicPhoneNumber() {
...
+    /**
+     * Prompt for getting the day of week from a user. Numbers the days of the week and returns the DayOfWeek
+     * object associated with the number.
+     * @return DayOfWeek - the day of the week (Ex: MONDAY, TUESDAY, etc.) the user chose or null if the user enters a
+     * number that is not associated with a DayOfWeek or doesn't enter a number.
+     */
+    public DayOfWeek showDayOfWeekPrompt(){
+        infoMessage("Select day of week: ");
+        ArrayList<DayOfWeek> dayOfWeeks = new ArrayList<>(Arrays.asList(DayOfWeek.MONDAY, DayOfWeek.TUESDAY,
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/ClinicScreenView.java]: As previously mentioned this arraylist, or LinkedHashMap should be gotten from the `dayOfWeekUtils` directly.

File: src/presenters/screenViews/ScreenView.java [hunk 28]
```diff
@@ -109,6 +110,18 @@ protected LocalDate showLocalDatePrompt() {
             return null;
         }
     }
+    protected LocalTime showLocalTimePrompt(){
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/ScreenView.java:116]: Add an empty line before the method.

File: src/presenters/screenViews/SecretaryScreenView.java [hunk 29]
```diff
@@ -154,12 +155,15 @@ public AppointmentTimeDetails bookAppointmentTimePrompt() {
     }
 
     /**
-     * Show error when appointment overlaps with absence or unavailability or another appointment.
+     * Show error when appointment overlaps with unavailability or another appointment.
      */
-    public void showAppointmentConflictError() {
-        errorMessage("Appointment booking error: time period unavailability.");
+    public void showAppointmentBookingError() {
+        errorMessage("Error Booking Appointment.");
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/SecretaryScreenView.java]: Please make this error message as explicit as the javadoc. I.e. error message too vague.

File: src/presenters/screenViews/SecretaryScreenView.java [hunk 30]
```diff
@@ -154,12 +155,15 @@ public AppointmentTimeDetails bookAppointmentTimePrompt() {
...
     }
 
+    public void showNoUserAppointmentsMessage(){
+        infoMessage("This user has no appointments.");
+    }
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/SecretaryScreenView.java:185]: Add an empty line after this method.

File: src/presenters/screenViews/SecretaryScreenView.java [hunk 31]
```diff
@@ -244,6 +248,18 @@ public void showDeleteNotAnIntegerError(String itemType) {
...
+    public void showCancelAppointmentSuccess(){
+        successMessage("Successfully canceled appointment.");
+    }
+
+    /**
+     * Message displayed when booking an appointment and showing clinic's availability on a given day, where there
+     * is no availability.
+     */
+    public void showNoAvailabilityError(ContactData userContact){
+        errorMessage("Can not book today as there is no availability.");
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/SecretaryScreenView.java]: Use correct English.

File: src/utilities/DayOfWeekUtils.java [hunk 32]
```diff
@@ -0,0 +1,25 @@
...
+    public String dayOfWeekToString(DayOfWeek dayOfWeek){
+        LinkedHashMap<DayOfWeek, String> dayMap = new LinkedHashMap<>() {{
+            put(DayOfWeek.MONDAY, "Monday");
+            put(DayOfWeek.TUESDAY, "Tuesday");
+            put(DayOfWeek.WEDNESDAY, "Wednesday");
+            put(DayOfWeek.THURSDAY, "Thursday");
+            put(DayOfWeek.FRIDAY, "Friday");
+            put(DayOfWeek.SATURDAY, "Saturday");
+            put(DayOfWeek.SUNDAY, "Sunday");
+        }};
```
Comment (ahmedkhalf) [inline src/utilities/DayOfWeekUtils.java:22]: As mentioned earlier, a method should be created to get this LinkedHashMap.

File: src/utilities/TimeUtils.java [hunk 33]
```diff
@@ -7,7 +7,7 @@
 import java.time.*;
```
Comment (ahmedkhalf) [inline src/utilities/TimeUtils.java]: Remove the `PHASE 2 FILE` comment at the start of this file.

File: src/utilities/TimeUtils.java [hunk 34]
```diff
@@ -23,8 +23,8 @@ public class TimeUtils {
      */
     public TimeBlock createTimeBlock(Integer year, Integer month, Integer day, Integer hour, Integer minute, Integer lenOfAppointment){
```
Comment (ahmedkhalf) [inline src/utilities/TimeUtils.java]: Line exceeds 120 chars.  Too many arguments: year, month, day can be replaced by local date. Hour, minute, can be replaced by local time.

File: src/useCases/DoctorManager.java [hunk 35]
```diff
@@ -1,10 +1,14 @@
...
 import dataBundles.DoctorData;
+import dataBundles.TimeBlockData;
 import database.DataMapperGateway;
 import database.Database;
 import entities.Doctor;
 
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+
 /**
```
Comment (ahmedkhalf) [inline src/useCases/DoctorManager.java]: On line 38 of this file it says pending implementation in phase 2???? (Sry intellij doesn't allow me to comment on non-edited lines).

File: src/useCases/AppointmentManager.java [hunk 36]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
                                            Integer lengthOfAppointments) {
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: Too many arguments: year, month, day can be replaced by local date. Hour, minute, can be replaced by local time.

File: src/useCases/AppointmentManager.java [hunk 37]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
+     * @param hour                 Integer - an integer value that represents an hour of a day.
+     * @param minute               Integer - an integer value that represents a minute of an hour.
      * @param lengthOfAppointments Integer - an integer value in minutes that represents the length of an appointment.
      * @return boolean - a boolean representing if the appointment successfully rescheduled.
      */
     public boolean rescheduleAppointment(AppointmentData appointmentData, Integer year, Integer month, Integer day,
-                                         Integer hour, Integer minute, Integer lengthOfAppointments){
-        Appointment appointment = appointmentDatabase.get(appointmentData.getAppointmentId());
-        appointmentDatabase.get(appointmentData.getAppointmentId()).setTimeBlock(null);
+                                         Integer hour, Integer minute, Integer lengthOfAppointments) {
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: Too many arguments: year, month, day can be replaced by local date. Hour, minute, can be replaced by local time.

File: src/useCases/AppointmentManager.java [hunk 38]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
-        ArrayList<TimeBlock> consideration = getTimeBlocksWithPatientAndDoctor(appointment.getDoctorId(),
-                appointment.getPatientId());
-        if (isNoTimeBlockConflictAppointment(consideration, proposedTime) &
-                isNoTimeBlockConflictAppointment(getSingleDayAvailability(appointmentData.getDoctorId(),
-                proposedTime.getStartTime().toLocalDate()), proposedTime) &
-                isNoTimeBlockConflictAbsence(appointmentData.getDoctorId(), proposedTime)){
-            appointmentDatabase.add(appointment);
+        Appointment selectedAppointment = appointmentDatabase.get(appointmentData.getAppointmentId());
+
+        database.getAppointmentDatabase().remove(appointmentData.getAppointmentId());
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: The appointment should only be removed inside the if statement.

File: src/useCases/AppointmentManager.java [hunk 39]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
+        DoctorData doctorData = new DoctorData(doctorDatabase.get(appointmentData.getDoctorId()));
+        if (isValidAppointment(doctorData, proposedTime)) {
+            appointmentDatabase.add(new Appointment(proposedTime, appointmentData.getDoctorId(), appointmentData.getPatientId()));
             return true;
         }
+        appointmentDatabase.add(selectedAppointment);
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: Once line 93 is moved inside the if statement, line 99 should be deleted.

File: src/useCases/AppointmentManager.java [hunk 40]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
-                isNoTimeBlockConflictAppointment(getSingleDayAvailability(appointmentData.getDoctorId(),
-                proposedTime.getStartTime().toLocalDate()), proposedTime) &
-                isNoTimeBlockConflictAbsence(appointmentData.getDoctorId(), proposedTime)){
-            appointmentDatabase.add(appointment);
+        Appointment selectedAppointment = appointmentDatabase.get(appointmentData.getAppointmentId());
+
+        database.getAppointmentDatabase().remove(appointmentData.getAppointmentId());
+        DoctorData doctorData = new DoctorData(doctorDatabase.get(appointmentData.getDoctorId()));
+        if (isValidAppointment(doctorData, proposedTime)) {
+            appointmentDatabase.add(new Appointment(proposedTime, appointmentData.getDoctorId(), appointmentData.getPatientId()));
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: line too long. (max 120 chars).

File: src/useCases/AppointmentManager.java [hunk 41]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
     /**
      * Gets all appointments related to a single doctor id.
+     *
      * @param doctorData DoctorData - data that represents the doctor entity.
      * @return ArrayList<AppointmentData> - ArrayList of AppointmentData which includes information of specific doctor
      * Appointments.
      */
-    public ArrayList<AppointmentData> getDoctorAppointments(DoctorData doctorData){
+    public ArrayList<AppointmentData> getDoctorAppointments(DoctorData doctorData) {
         return getAppointments().stream()
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java:118]: if you are going to have a `getAllPatientAppointments` method, then if it's fair to say you should have `getAllDoctorAppointments` method here. I.e. be consistent.

File: src/useCases/AppointmentManager.java [hunk 42]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
-                .map(Appointment::getTimeBlock)
                 .collect(Collectors.toCollection(ArrayList::new));
     }
 
-    private ArrayList<Appointment> getAllPatientAppointments(Integer patientId){
-        return getAppointments().stream()
-                .filter(x -> x.getPatientId().equals(patientId))
-                .collect(Collectors.toCollection(ArrayList::new));
+    private boolean isValidAppointment(DoctorData doctorData, TimeBlock timeBlock) {
+        return doesNotOverlapWithDoctorsAppointments(timeBlock, doctorData) && strictlyOverlapsWithClinicHours(timeBlock);
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: Line over 120 chars.

File: src/useCases/AppointmentManager.java [hunk 43]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
+    private boolean strictlyWithinHours(UniversalTimeBlock timeBlock, UniversalTimeBlock timeBlock2) {
+        return isWithinHours(timeBlock, timeBlock2.startTime()) && isWithinHours(timeBlock, timeBlock2.endTime());
     }
-
-    private void removeAppointmentConflictAvailability(DoctorData doctorData, Availability availability){
-        getDoctorAppointments(doctorData).stream()
-                .filter(x-> overlapDay(x, availability))
-                .filter(x -> overlapTime(x, availability))
-                .forEach(this::removeAppointment);
+    private boolean isWithinDate(LocalDate time1, LocalDate time2, UniversalTimeBlockWithDay timeBlock) {
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: This method is not being used? Any reason why?

File: src/useCases/AppointmentManager.java [hunk 44]
```diff
@@ -53,357 +55,151 @@ public AppointmentData bookAppointment(PatientData patientData, DoctorData docto
...
-                    .filter(x -> x.getStartTime().getDayOfYear() ==
-                            availabilityTimeBlock.getStartTime().getDayOfYear())
-                    .filter(x -> calculateNoConflictTime(x, startTime, parsedAvailability))
-                    .close();
-            if (startTime.get(0) != availabilityTimeBlock.getEndTime()){
-                parsedAvailability.add(new TimeBlock(startTime.remove(0), availabilityTimeBlock.getEndTime()));
-            }
-        }
-        return parsedAvailability;
+    private boolean doesNotOverlapWithDoctorsAppointments(UniversalTimeBlockWithDay timeBlock, DoctorData doctorData) {
```
Comment (ahmedkhalf) [inline src/useCases/AppointmentManager.java]: Im not quite sure, how line 174 until the end of the file work. I know they are private methods but maybe a javadoc with a little explanation will help. If the tests cover all edge cases though, then I will trust this code.

File: test/AppointmentManagerTests.java [hunk 45]
```diff
@@ -66,96 +66,205 @@ public void testBookAppointment() {
...
-//                .get(validAppointment3.getAppointmentId()).getId());
-//        AppointmentData invalidAppointment6 = new AppointmentManager(originalDatabase).bookAppointment(patientData,
-//                doctorData, 2022, 12, 5, 1, 0, 120);
-//
-//        assertNull("an appointment should be returning null when booking at an invalid time",
-//                invalidAppointment6);
-//    }
+        originalDatabase.setClinic(new Clinic("", "", "", "",
+                new ArrayList<>(List.of(new Availability(DayOfWeek.of(1), LocalTime.of(8, 30),
+                        LocalTime.of(17, 0))))));
```
Comment (ahmedkhalf) [inline test/AppointmentManagerTests.java]: Why are you reeating the same code for every test.... Use helper methods.

File: test/AppointmentManagerTests.java [hunk 46]
```diff
@@ -66,96 +66,205 @@ public void testBookAppointment() {
...
 
+        DoctorData doctorData = new DoctorManager(originalDatabase).createDoctor("test1", "test1");
```
Comment (ahmedkhalf) [inline test/AppointmentManagerTests.java]: I refuse to review the tests any further. This is not sane.

File: src/controllers/SecretaryLoadedPatientController.java [hunk 47]
```diff
@@ -121,21 +124,26 @@ private Command bookAppointment() {
...
             } else {
                 appointmentManager.removeAppointment(data.get(index));
+                secretaryScreenView.showCancelAppointmentSuccess();
             }
         };
     }
-
     private Command rescheduleAppointment() {
         return (x) -> {
             ArrayList<AppointmentData> appointments = appointmentManager.getPatientAppointments(patientData);
```
Comment (ahmedkhalf) [inline src/controllers/SecretaryLoadedPatientController.java:200]: `showNoUserAppointmentsMessage` should be shown if patient has no appointments.

File: src/controllers/SecretaryLoadedPatientController.java [hunk 48]
```diff
@@ -155,20 +163,35 @@ private Command rescheduleAppointment() {
         };
```
Comment (ahmedkhalf) [inline src/controllers/SecretaryLoadedPatientController.java:240]: On line 159, why is there no error handling? What happens after rescheduleAppointment returns false?

File: src/controllers/SecretaryLoadedPatientController.java [hunk 49]
```diff
@@ -155,20 +163,35 @@ private Command rescheduleAppointment() {
...
+        ContactData doctorContact = contactManager.getContactData(doctorData);
+
+        ArrayList<AppointmentData> appointments = appointmentManager.getSingleDayAppointment(
+                doctorData, LocalDate.of(date.getYear(), date.getMonthValue(), date.getDayOfMonth()));
+
+
+        AvailabilityData availabilityData = appointmentManager.getAvailabilityFromDayOfWeek(date.getDayOfWeek());
+        if (availabilityData == null){
+            secretaryScreenView.showNoAvailabilityError(doctorContact);
+        }else {
```
Comment (ahmedkhalf) [inline src/controllers/SecretaryLoadedPatientController.java]: a presenter and a check for no appointments should also be added.

File: src/controllers/PatientController.java [hunk 50]
```diff
@@ -30,28 +36,26 @@ public PatientController(Context context, PatientData patientData) {
...
         prescriptionController.AllCommands().forEach((x, y) -> commands.put("view " + x, y));
         commands.putAll(super.AllCommands());
         return commands;
     }
 
-    /* PENDING IMPLEMENTATION IN PHASE 2
     private Command ViewAppointments() {
-        AppointmentManager appointmentManager = new AppointmentManager(getDatabase());
         return (x) -> {
             ArrayList<AppointmentData> appointments = appointmentManager.getPatientAppointments(patientData);
```
Comment (ahmedkhalf) [inline src/controllers/PatientController.java:56]: a check and presenter method should be added, for when there are no appointments.

File: src/controllers/DoctorController.java [hunk 51]
```diff
@@ -68,63 +68,23 @@ private Command LoadPatient() {
...
-            Integer deleteInteger = doctorScreenView.deleteAbsencePrompt(new ContactManager(getDatabase())
-                    .getContactData(doctorData), doctorData.getAbsence().stream()
-                    .map(TimeBlockData::new)
-                    .collect(Collectors.toCollection(ArrayList::new)));
-            new AppointmentManager(getDatabase()).deleteAbsence(doctorData, doctorData.getAbsence().get(deleteInteger));
-        };
+    private Command ViewAllDoctorAppointments() {
+        return (x) -> doctorScreenView.viewAppointments(new AppointmentManager(getDatabase())
+                .getDoctorAppointments(doctorData));
     }
```
Comment (ahmedkhalf) [inline src/controllers/DoctorController.java:96]: for all three view methods from line 71 to 89 a check and no appointments message should be added.

File: src/controllers/ClinicController.java [hunk 52]
```diff
@@ -98,4 +104,29 @@ private Command ChangeClinicAddress() {
         };
     }
 
+    private Command ChangeClinicHours(){
+        return (x) -> {
+            clinicScreenView.showClinicHours(clinicManager.clinicData());
+            AvailabilityDetails availabilityDetails = clinicScreenView.showChangeClinicHoursPrompt();
+            if (availabilityDetails == null || availabilityDetails.getStartTime().isAfter(availabilityDetails.getEndTime())){
```
Comment (ahmedkhalf) [inline src/controllers/ClinicController.java]: Line exceeds 120 chars.

File: src/controllers/ClinicController.java [hunk 53]
```diff
@@ -98,4 +104,29 @@ private Command ChangeClinicAddress() {
...
+                clinicScreenView.showEnteredInvalidAvailabilityInfoError();
+                return;
+            }
+            clinicManager.changeClinicHours(availabilityDetails.getDayOfWeek(), availabilityDetails.getStartTime(), availabilityDetails.getEndTime());
```
Comment (ahmedkhalf) [inline src/controllers/ClinicController.java]: Line exceeds 120 chars.

File: src/presenters/screenViews/ClinicScreenView.java [hunk 54]
```diff
@@ -90,4 +103,78 @@ public void showSuccessfullyChangedClinicPhoneNumber() {
...
+     */
+    public void showSuccessfullyChangedAvailability(){
+        successMessage("Successfully changed the clinic's availability.");
+    }
+
+    /**
+     * Shows an error message when a user incorrectly enters a field when changing clinic hours.
+     */
+    public void showEnteredInvalidAvailabilityInfoError(){
+        errorMessage("Invalid entry.");
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/ClinicScreenView.java]: Please be more descriptive.

File: src/presenters/screenViews/ClinicScreenView.java [hunk 55]
```diff
@@ -90,4 +103,78 @@ public void showSuccessfullyChangedClinicPhoneNumber() {
...
+    }
+
+    /**
+     * Shows an error when a user selects an invalid day of week when removing availability.
+     */
+    public void showInvalidSelectionError(){
+        errorMessage("Invalid selection.");
```
Comment (ahmedkhalf) [inline src/presenters/screenViews/ClinicScreenView.java]: Please be more descriptive.

**Review by ahmedkhalf — CHANGES_REQUESTED** [review]
Overall great code. Very few minor mistakes are to be fixed. Good job guys.

File: src/entities/UniversalTimeBlock.java (hunk 3 above)
Comment (ShlomoAlon) [reply src/entities/UniversalTimeBlock.java]: It causes annoying naming conflicts. Not sure if its worth fixing. the other option was getStartTimeLocalTime but this is really long

File: src/entities/UniversalTimeBlock.java (hunk 4 above)
Comment (ShlomoAlon) [reply src/entities/UniversalTimeBlock.java]: naming conflicts as above

File: src/entities/UniversalTimeBlock.java (hunk 5 above)
Comment (ShlomoAlon) [reply src/entities/UniversalTimeBlock.java]: naming conflicts

File: src/entities/TimeBlock.java (hunk 14 above)
Comment (ShlomoAlon) [reply src/entities/TimeBlock.java]: its in LocalDateTime not LocalTime

File: src/controllers/SecretaryLoadedPatientController.java [hunk 56]
```diff
@@ -150,27 +144,74 @@ private Command RescheduleAppointment() {
             } else {
                 LocalDate day = secretaryScreenView.bookAppointmentDayPrompt();
                 AppointmentTimeDetails time = secretaryScreenView.bookAppointmentTimePrompt();
-                appointmentManager.rescheduleAppointment(
-                        appointments.get(index), day.getYear(), day.getMonthValue(),
-                        day.getDayOfMonth(), time.time().getHour(), time.time().getMinute(), time.length());
+                LocalDateTime startTime = LocalDateTime.of(day.getYear(), day.getMonthValue(), day.getDayOfMonth(),
```
Comment (ahmedkhalf) [inline src/controllers/SecretaryLoadedPatientController.java]: What happens when day is null? This code smells like null pointer errors. What happens when time is null? Again null pointer errors. Furthermore, if the call to  `getPatientAppointments` returns an empty list a relevant presenter method NoUserAppointmentsMessage should be displayed and the command should not continue.

File: src/controllers/PatientController.java (hunk 50 above)
Comment (ahmedkhalf) [inline src/controllers/PatientController.java:56]: A check and presenter method should be added, for when there are no appointments.

File: src/controllers/DoctorController.java [hunk 57]
```diff
@@ -68,63 +64,26 @@ private Command LoadPatient() {
...
-    private Command DeleteAbsence() {
+    private Command ViewSchedule() {
         return (x) -> {
-            Integer deleteInteger = doctorScreenView.deleteAbsencePrompt(new ContactManager(getDatabase())
-                    .getContactData(doctorData), doctorData.getAbsence().stream()
-                    .map(TimeBlockData::new)
-                    .collect(Collectors.toCollection(ArrayList::new)));
-            new AppointmentManager(getDatabase()).deleteAbsence(doctorData, doctorData.getAbsence().get(deleteInteger));
+            LocalDate viewDate = doctorScreenView.viewSchedulePrompt();
+            doctorScreenView.viewAppointments(new AppointmentManager(getDatabase()).getSingleDayAppointment(doctorData, viewDate));
```
Comment (ahmedkhalf) [inline src/controllers/DoctorController.java]: A check and no appointments message should be added.

File: src/controllers/DoctorController.java [hunk 58]
```diff
@@ -68,63 +64,26 @@ private Command LoadPatient() {
...
-            new AppointmentManager(getDatabase()).addAbsence(doctorData, new TimeUtils()
-                    .createLocalDateTime(absenceData.get(0), absenceData.get(1),
-                    absenceData.get(2), 0, 0), new TimeUtils().createLocalDataTime(absenceData.get(0),
-                    absenceData.get(1), absenceData.get(2) + absenceData.get(3), 0, 0));
-        };
-    } */
-
+    private Command ViewAllDoctorAppointments() {
+        return (x) -> doctorScreenView.viewAppointments(new AppointmentManager(getDatabase())
+                .getDoctorAppointments(doctorData));
```
Comment (ahmedkhalf) [inline src/controllers/DoctorController.java]: A check and no appointments message should be added.

**Review by ahmedkhalf — CHANGES_REQUESTED** [review]

File: src/controllers/DoctorController.java (hunk 58 above)
Comment (DanielDervishi) [reply src/controllers/DoctorController.java]: This command should have been removed in the first place.

**Review by ahmedkhalf — APPROVED** [review]

**Review by Hepaul7 — APPROVED** [review]
Looks good to me!
